File: services/ledger/src/ledger.ts
import {
  AssetSymbol,
  AccountType,
  TransactionType,
  EntryDirection,
  JournalEntry,
  LedgerTransaction,
  UserBalance,
  Decimal,
  FinancialInvariantError,
  InsufficientBalanceError,
  Logger
} from '@syncnode/common';
import { db } from '@syncnode/database';

export interface EntrySpec {
  userId?: string;
  accountType: AccountType;
  asset: AssetSymbol;
  direction: EntryDirection;
  amount: string;
}

export class LedgerService {
  private readonly logger = new Logger('LedgerService');

  /**
   * Fetch or initialize an account balance in the ledger.
   */
  public getOrCreateAccount(userId: string | undefined, type: AccountType, asset: AssetSymbol) {
    const accountId = db.getAccountId(userId, type, asset);
    let account = db.accounts.get(accountId);
    if (!account) {
      account = {
        id: accountId,
        userId,
        type,
        asset,
        balance: '0'
      };
      db.accounts.set(accountId, account);
    }
    return account;
  }

  /**
   * Execute an atomic, multi-entry double-entry ledger transaction.
   * Enforces mathematical invariant: Debits == Credits per asset.
   */
  public recordTransaction(
    type: TransactionType,
    referenceId: string,
    idempotencyKey: string,
    description: string,
    entriesSpec: EntrySpec[]
  ): LedgerTransaction {
    // 1. Idempotency Check
    if (db.idempotencyKeys.has(idempotencyKey)) {
      const existing = Array.from(db.ledgerTransactions.values()).find(
        (t) => t.idempotencyKey === idempotencyKey
      );
      if (existing) {
        this.logger.info(`Idempotent transaction replayed: ${idempotencyKey}`);
        return existing;
      }
    }

    if (entriesSpec.length < 2) {
      throw new FinancialInvariantError('A double-entry transaction must have at least 2 entries');
    }

    // 2. Validate Debits == Credits per asset
    const assetTotals = new Map<AssetSymbol, { debit: Decimal; credit: Decimal }>();

    for (const spec of entriesSpec) {
      const amount = new Decimal(spec.amount);
      if (!amount.isPositive()) {
        throw new FinancialInvariantError(`Transaction entry amount must be positive, got ${spec.amount}`);
      }

      if (!assetTotals.has(spec.asset)) {
        assetTotals.set(spec.asset, { debit: Decimal.ZERO, credit: Decimal.ZERO });
      }
      const totals = assetTotals.get(spec.asset)!;

      if (spec.direction === EntryDirection.DEBIT) {
        totals.debit = totals.debit.plus(amount);
      } else {
        totals.credit = totals.credit.plus(amount);
      }
    }

    for (const [asset, totals] of assetTotals.entries()) {
      if (!totals.debit.eq(totals.credit)) {
        throw new FinancialInvariantError(
          `Unbalanced ledger entry for asset ${asset}: Debit=${totals.debit.toString()} != Credit=${totals.credit.toString()}`
        );
      }
    }

    // 3. Pre-flight check for liability accounts to ensure no negative balances
    for (const spec of entriesSpec) {
      // In standard exchange accounting:
      // For Liability accounts (User balances):
      // DEBIT decreases user liability (reduces balance)
      // CREDIT increases user liability (increases balance)
      // If debiting a user account, ensure they have sufficient balance.
      if (
        spec.direction === EntryDirection.DEBIT &&
        [
          AccountType.USER_AVAILABLE,
          AccountType.USER_LOCKED,
          AccountType.USER_WITHDRAWAL_PENDING,
          AccountType.USER_P2P_ESCROW
        ].includes(spec.accountType)
      ) {
        const account = this.getOrCreateAccount(spec.userId, spec.accountType, spec.asset);
        const current = new Decimal(account.balance);
        const debitAmt = new Decimal(spec.amount);
        if (current.lt(debitAmt)) {
          throw new InsufficientBalanceError(spec.asset, debitAmt.toString(), current.toString());
        }
      }
    }

    // 4. Apply State Changes Atomically
    const txId = `tx_${Date.now()}_${Math.random().toString(36).substring(2, 9)}`;
    const journalEntries: JournalEntry[] = [];

    for (const spec of entriesSpec) {
      const account = this.getOrCreateAccount(spec.userId, spec.accountType, spec.asset);
      const current = new Decimal(account.balance);
      const amount = new Decimal(spec.amount);

      let newBalance: Decimal;
      // Asset accounts (Vaults): DEBIT increases, CREDIT decreases
      if (
        spec.accountType === AccountType.EXCHANGE_HOT_WALLET ||
        spec.accountType === AccountType.EXCHANGE_COLD_STORAGE
      ) {
        if (spec.direction === EntryDirection.DEBIT) {
          newBalance = current.plus(amount);
        } else {
          newBalance = current.minus(amount);
        }
      } else {
        // Liability & Revenue accounts: CREDIT increases, DEBIT decreases
        if (spec.direction === EntryDirection.CREDIT) {
          newBalance = current.plus(amount);
        } else {
          newBalance = current.minus(amount);
        }
      }

      account.balance = newBalance.toString();

      const entry: JournalEntry = {
        id: `je_${Date.now()}_${Math.random().toString(36).substring(2, 9)}`,
        accountId: account.id,
        userId: spec.userId,
        accountType: spec.accountType,
        asset: spec.asset,
        direction: spec.direction,
        amount: spec.amount,
        createdAt: Date.now()
      };

      db.journalEntries.set(entry.id, entry);
      journalEntries.push(entry);
    }

    const tx: LedgerTransaction = {
      id: txId,
      type,
      referenceId,
      idempotencyKey,
      description,
      entries: journalEntries,
      createdAt: Date.now()
    };

    db.ledgerTransactions.set(txId, tx);
    db.idempotencyKeys.add(idempotencyKey);

    return tx;
  }

  /**
   * Retrieve structured multi-asset balances for a user.
   */
  public getUserBalances(userId: string): UserBalance[] {
    const assets = Object.values(AssetSymbol);
    const result: UserBalance[] = [];

    for (const asset of assets) {
      const availAcc = this.getOrCreateAccount(userId, AccountType.USER_AVAILABLE, asset);
      const lockedAcc = this.getOrCreateAccount(userId, AccountType.USER_LOCKED, asset);
      const pendAcc = this.getOrCreateAccount(userId, AccountType.USER_WITHDRAWAL_PENDING, asset);
      const p2pAcc = this.getOrCreateAccount(userId, AccountType.USER_P2P_ESCROW, asset);

      const available = new Decimal(availAcc.balance);
      const locked = new Decimal(lockedAcc.balance);
      const pending = new Decimal(pendAcc.balance);
      const p2p = new Decimal(p2pAcc.balance);
      const total = available.plus(locked).plus(pending).plus(p2p);

      result.push({
        asset,
        available: available.toString(),
        locked: locked.toString(),
        pendingWithdrawal: pending.toString(),
        p2pEscrow: p2p.toString(),
        total: total.toString()
      });
    }

    return result;
  }

  /**
   * Get single asset balance for a user.
   */
  public getUserAssetBalance(userId: string, asset: AssetSymbol): UserBalance {
    const availAcc = this.getOrCreateAccount(userId, AccountType.USER_AVAILABLE, asset);
    const lockedAcc = this.getOrCreateAccount(userId, AccountType.USER_LOCKED, asset);
    const pendAcc = this.getOrCreateAccount(userId, AccountType.USER_WITHDRAWAL_PENDING, asset);
    const p2pAcc = this.getOrCreateAccount(userId, AccountType.USER_P2P_ESCROW, asset);

    const available = new Decimal(availAcc.balance);
    const locked = new Decimal(lockedAcc.balance);
    const pending = new Decimal(pendAcc.balance);
    const p2p = new Decimal(p2pAcc.balance);
    const total = available.plus(locked).plus(pending).plus(p2p);

    return {
      asset,
      available: available.toString(),
      locked: locked.toString(),
      pendingWithdrawal: pending.toString(),
      p2pEscrow: p2p.toString(),
      total: total.toString()
    };
  }

  /**
   * Proof of Solvency & Reserves Mathematical Audit.
   * Compares Total Vault Assets vs Total Customer Liabilities & Exchange Equity.
   */
  public performProofOfReservesAudit(): {
    isSolvent: boolean;
    assets: Record<string, { totalAssets: string; totalLiabilities: string; surplus: string; ratio: string }>;
    timestamp: number;
  } {
    const assets = Object.values(AssetSymbol);
    const report: Record<string, { totalAssets: string; totalLiabilities: string; surplus: string; ratio: string }> = {};
    let overallSolvent = true;

    for (const asset of assets) {
      let vaultAssets = Decimal.ZERO;
      let userLiabilities = Decimal.ZERO;

      for (const account of db.accounts.values()) {
        if (account.asset !== asset) continue;
        const bal = new Decimal(account.balance);

        if (
          account.type === AccountType.EXCHANGE_HOT_WALLET ||
          account.type === AccountType.EXCHANGE_COLD_STORAGE
        ) {
          vaultAssets = vaultAssets.plus(bal);
        } else if (
          account.type === AccountType.USER_AVAILABLE ||
          account.type === AccountType.USER_LOCKED ||
          account.type === AccountType.USER_WITHDRAWAL_PENDING ||
          account.type === AccountType.USER_P2P_ESCROW
        ) {
          userLiabilities = userLiabilities.plus(bal);
        }
      }

      const surplus = vaultAssets.minus(userLiabilities);
      const isSolventForAsset = vaultAssets.gte(userLiabilities);
      if (!isSolventForAsset) overallSolvent = false;

      const ratio = userLiabilities.isZero()
        ? '100.00%'
        : `${vaultAssets.dividedBy(userLiabilities).times(100).toFixed(2)}%`;

      report[asset] = {
        totalAssets: vaultAssets.toString(),
        totalLiabilities: userLiabilities.toString(),
        surplus: surplus.toString(),
        ratio
      };
    }

    return {
      isSolvent: overallSolvent,
      assets: report,
      timestamp: Date.now()
    };
  }

  /**
   * Full treasury summary per asset, broken down by account type.
   * All figures originate from the double-entry ledger (authoritative).
   * Values are expressed in native asset units only - no fiat valuation is
   * fabricated without a trusted pricing source.
   */
  public getTreasurySummary(): {
    assets: Record<string, {
      hotWallet: string;
      coldStorage: string;
      totalExchangeAssets: string;
      liabilities: {
        available: string;
        locked: string;
        pendingWithdrawal: string;
        p2pEscrow: string;
        total: string;
      };
      revenue: { tradingFees: string; withdrawalFees: string };
      reserveRatio: string;
      isSolvent: boolean;
      withdrawalCapacityRatio: string;
    }>;
    timestamp: number;
  } {
    const assets = Object.values(AssetSymbol);
    const report: Record<string, any> = {};

    for (const asset of assets) {
      let hotWallet = Decimal.ZERO;
      let coldStorage = Decimal.ZERO;
      let available = Decimal.ZERO;
      let locked = Decimal.ZERO;
      let pendingWithdrawal = Decimal.ZERO;
      let p2pEscrow = Decimal.ZERO;
      let tradingFees = Decimal.ZERO;
      let withdrawalFees = Decimal.ZERO;

      for (const account of db.accounts.values()) {
        if (account.asset !== asset) continue;
        const bal = new Decimal(account.balance);

        switch (account.type) {
          case AccountType.EXCHANGE_HOT_WALLET:
            hotWallet = hotWallet.plus(bal);
            break;
          case AccountType.EXCHANGE_COLD_STORAGE:
            coldStorage = coldStorage.plus(bal);
            break;
          case AccountType.USER_AVAILABLE:
            available = available.plus(bal);
            break;
          case AccountType.USER_LOCKED:
            locked = locked.plus(bal);
            break;
          case AccountType.USER_WITHDRAWAL_PENDING:
            pendingWithdrawal = pendingWithdrawal.plus(bal);
            break;
          case AccountType.USER_P2P_ESCROW:
            p2pEscrow = p2pEscrow.plus(bal);
            break;
          case AccountType.TRADING_FEES:
            tradingFees = tradingFees.plus(bal);
            break;
          case AccountType.WITHDRAWAL_FEES:
            withdrawalFees = withdrawalFees.plus(bal);
            break;
          default:
            break;
        }
      }

      const totalExchangeAssets = hotWallet.plus(coldStorage);
      const totalLiabilities = available.plus(locked).plus(pendingWithdrawal).plus(p2pEscrow);

      // Withdrawal capacity: liquid hot-wallet coverage over outstanding
      // pending withdrawals plus a same-asset buffer reference. Expressed as
      // ratio of hot wallet to pending withdrawals when withdrawals exist.
      const withdrawalCapacityRatio = pendingWithdrawal.isZero()
        ? 'INFINITE'
        : `${hotWallet.dividedBy(pendingWithdrawal).toFixed(4)}x`;

      report[asset] = {
        hotWallet: hotWallet.toString(),
        coldStorage: coldStorage.toString(),
        totalExchangeAssets: totalExchangeAssets.toString(),
        liabilities: {
          available: available.toString(),
          locked: locked.toString(),
          pendingWithdrawal: pendingWithdrawal.toString(),
          p2pEscrow: p2pEscrow.toString(),
          total: totalLiabilities.toString()
        },
        revenue: {
          tradingFees: tradingFees.toString(),
          withdrawalFees: withdrawalFees.toString()
        },
        reserveRatio: totalLiabilities.isZero()
          ? '100.00%'
          : `${totalExchangeAssets.dividedBy(totalLiabilities).times(100).toFixed(2)}%`,
        isSolvent: totalExchangeAssets.gte(totalLiabilities),
        withdrawalCapacityRatio
      };
    }

    return { assets: report, timestamp: Date.now() };
  }
}

export const ledgerService = new LedgerService();

File: syncnode/services/ledger.py
import time
import uuid
from typing import List, Dict, Any, Optional
from syncnode.common.types import AssetSymbol, AccountType
from syncnode.common.errors import FinancialInvariantError, InsufficientFundsError
from syncnode.common.decimal_util import (
    to_decimal,
    format_decimal,
    add_decimals,
    sub_decimals,
    gt_decimal,
    gte_decimal,
    lt_decimal,
    eq_decimal
)
from syncnode.database.db import db
from syncnode.database.supabase_client import supabase_client
from syncnode.common.logger import Logger

logger = Logger("LedgerService")


class LedgerService:
    def get_or_create_account(self, account_type: AccountType, asset: AssetSymbol, user_id: Optional[str] = None) -> Dict[str, Any]:
        acc_id = f"acc_{user_id or 'sys'}_{account_type.value}_{asset.value}"
        if acc_id not in db.accounts:
            acc_data = {
                "id": acc_id,
                "user_id": user_id,
                "type": account_type.value if hasattr(account_type, "value") else str(account_type),
                "asset": asset.value if hasattr(asset, "value") else str(asset),
                "balance": "0.00000000",
                "created_at": int(time.time() * 1000),
                "updated_at": int(time.time() * 1000)
            }
            db.accounts[acc_id] = acc_data
            supabase_client.queue_upsert("accounts", acc_data)
        return db.accounts[acc_id]

    def record_transaction(self, idempotency_key: str, description: str, entries: List[Dict[str, Any]], metadata: Dict[str, Any] = None) -> Dict[str, Any]:
        if idempotency_key in db.idempotency_keys:
            # Return existing transaction if duplicate request
            for tx in db.ledger_transactions.values():
                if tx.get("idempotency_key") == idempotency_key:
                    return tx

        # Group entries by asset and verify that sum(debits) == sum(credits)
        asset_sums: Dict[str, Dict[str, Any]] = {}
        for entry in entries:
            asset = entry["asset"]
            asset_str = asset.value if hasattr(asset, "value") else str(asset)
            if asset_str not in asset_sums:
                asset_sums[asset_str] = {"debits": to_decimal(0), "credits": to_decimal(0)}
            asset_sums[asset_str]["debits"] += to_decimal(entry.get("debit", 0))
            asset_sums[asset_str]["credits"] += to_decimal(entry.get("credit", 0))

        for asset_name, sums in asset_sums.items():
            if sums["debits"] != sums["credits"]:
                raise FinancialInvariantError(
                    f"Double-entry invariant violated for {asset_name}: total debits {sums['debits']} != total credits {sums['credits']}"
                )

        # Pre-verify that debiting accounts have sufficient funds
        for entry in entries:
            debit_amt = to_decimal(entry.get("debit", 0))
            if debit_amt > 0:
                acc_id = entry["account_id"]
                current_acc = db.accounts.get(acc_id)
                if current_acc and current_acc["type"] != AccountType.EXCHANGE_VAULT:
                    current_bal = to_decimal(current_acc["balance"])
                    if current_bal < debit_amt:
                        raise InsufficientFundsError(f"Insufficient funds in account {acc_id}: balance {current_bal} < debit {debit_amt}")

        # Execute balance updates
        tx_id = f"tx_{int(time.time() * 1000)}_{uuid.uuid4().hex[:8]}"
        recorded_entries = []

        for entry in entries:
            acc_id = entry["account_id"]
            debit_amt = to_decimal(entry.get("debit", 0))
            credit_amt = to_decimal(entry.get("credit", 0))
            account = db.accounts.get(acc_id)
            if not account:
                raise FinancialInvariantError(f"Account {acc_id} does not exist")

            bal = to_decimal(account["balance"])
            if account["type"] == AccountType.EXCHANGE_VAULT:
                # Vault is an asset account: debits increase vault assets, credits decrease
                new_bal = bal + debit_amt - credit_amt
            else:
                # User / liability accounts: credits increase, debits decrease
                new_bal = bal + credit_amt - debit_amt

            if new_bal < 0:
                raise FinancialInvariantError(f"Negative balance invariant violated for account {acc_id}: {new_bal}")

            account["balance"] = format_decimal(new_bal)
            account["updated_at"] = int(time.time() * 1000)

            # Persist updated balance to Supabase PostgreSQL table
            acc_type_val = account["type"].value if hasattr(account["type"], "value") else str(account["type"])
            acc_asset_val = account["asset"].value if hasattr(account["asset"], "value") else str(account["asset"])
            supabase_client.queue_upsert("accounts", {
                "id": acc_id,
                "user_id": account.get("user_id"),
                "type": acc_type_val,
                "asset": acc_asset_val,
                "balance": format_decimal(new_bal),
                "created_at": account.get("created_at", int(time.time() * 1000)),
                "updated_at": account["updated_at"]
            })

            j_id = f"je_{uuid.uuid4().hex[:12]}"
            journal_entry = {
                "id": j_id,
                "transaction_id": tx_id,
                "account_id": acc_id,
                "asset": entry["asset"].value if hasattr(entry["asset"], "value") else str(entry["asset"]),
                "debit": format_decimal(debit_amt),
                "credit": format_decimal(credit_amt),
                "created_at": int(time.time() * 1000)
            }
            db.journal_entries[j_id] = journal_entry
            recorded_entries.append(journal_entry)
            supabase_client.queue_upsert("journal_entries", journal_entry)

        tx_record = {
            "id": tx_id,
            "idempotency_key": idempotency_key,
            "description": description,
            "entries": recorded_entries,
            "metadata": metadata or {},
            "created_at": int(time.time() * 1000)
        }
        db.ledger_transactions[tx_id] = tx_record
        db.idempotency_keys.add(idempotency_key)
        supabase_client.queue_upsert("ledger_transactions", tx_record)
        return tx_record

    def get_user_asset_balance(self, user_id: str, asset: AssetSymbol) -> Dict[str, str]:
        avail_acc = self.get_or_create_account(AccountType.USER_AVAILABLE, asset, user_id)
        locked_acc = self.get_or_create_account(AccountType.USER_LOCKED, asset, user_id)
        available = avail_acc["balance"]
        locked = locked_acc["balance"]
        total = add_decimals(available, locked)
        return {
            "available": available,
            "locked": locked,
            "total": total
        }

    def get_user_balances(self, user_id: str) -> List[Dict[str, Any]]:
        balances = []
        for asset in AssetSymbol:
            bal_info = self.get_user_asset_balance(user_id, asset)
            balances.append({
                "asset": asset.value,
                "available": bal_info["available"],
                "locked": bal_info["locked"],
                "total": bal_info["total"]
            })
        return balances

    def lock_user_funds(self, user_id: str, asset: AssetSymbol, amount: str, reason: str) -> Dict[str, Any]:
        avail_acc = self.get_or_create_account(AccountType.USER_AVAILABLE, asset, user_id)
        locked_acc = self.get_or_create_account(AccountType.USER_LOCKED, asset, user_id)
        idemp = f"lock_{user_id}_{asset.value}_{int(time.time()*1000)}_{uuid.uuid4().hex[:6]}"
        return self.record_transaction(
            idempotency_key=idemp,
            description=f"Lock funds: {reason}",
            entries=[
                {"account_id": avail_acc["id"], "asset": asset, "debit": amount, "credit": "0.00000000"},
                {"account_id": locked_acc["id"], "asset": asset, "debit": "0.00000000", "credit": amount}
            ]
        )

    def unlock_user_funds(self, user_id: str, asset: AssetSymbol, amount: str, reason: str) -> Dict[str, Any]:
        avail_acc = self.get_or_create_account(AccountType.USER_AVAILABLE, asset, user_id)
        locked_acc = self.get_or_create_account(AccountType.USER_LOCKED, asset, user_id)
        idemp = f"unlock_{user_id}_{asset.value}_{int(time.time()*1000)}_{uuid.uuid4().hex[:6]}"
        return self.record_transaction(
            idempotency_key=idemp,
            description=f"Unlock funds: {reason}",
            entries=[
                {"account_id": locked_acc["id"], "asset": asset, "debit": amount, "credit": "0.00000000"},
                {"account_id": avail_acc["id"], "asset": asset, "debit": "0.00000000", "credit": amount}
            ]
        )

    def verify_proof_of_reserves(self) -> Dict[str, Any]:
        solvency_report = {}
        assets_audit = {}
        all_solvent = True
        now = int(time.time() * 1000)

        for asset in AssetSymbol:
            vault_acc = self.get_or_create_account(AccountType.EXCHANGE_VAULT, asset)
            vault_balance = to_decimal(vault_acc["balance"])

            # Sum all customer obligations (Available + Locked + P2P Escrow)
            customer_obligations = to_decimal(0)
            for acc in db.accounts.values():
                if acc["asset"] == asset and acc["type"] in [AccountType.USER_AVAILABLE, AccountType.USER_LOCKED, AccountType.P2P_ESCROW]:
                    customer_obligations += to_decimal(acc["balance"])

            surplus_dec = vault_balance - customer_obligations
            ratio_dec = (vault_balance / customer_obligations) if customer_obligations > 0 else to_decimal(1)
            is_asset_solvent = vault_balance >= customer_obligations
            if not is_asset_solvent:
                all_solvent = False

            solvency_report[asset.value] = {
                "vault_reserve": format_decimal(vault_balance),
                "customer_liabilities": format_decimal(customer_obligations),
                "collateral_ratio_pct": format_decimal(ratio_dec * 100, 2),
                "is_solvent": is_asset_solvent
            }

            assets_audit[asset.value] = {
                "totalAssets": format_decimal(vault_balance),
                "totalLiabilities": format_decimal(customer_obligations),
                "surplus": format_decimal(surplus_dec),
                "ratio": format_decimal(ratio_dec, 4)
            }

        audit_payload = {
            "isSolvent": all_solvent,
            "assets": assets_audit,
            "timestamp": now
        }

        return {
            "proofOfReserves": solvency_report,
            "audit": audit_payload,
            "isSolvent": all_solvent,
            "timestamp": now,
            **solvency_report
        }

    def get_detailed_treasury_summary(self) -> Dict[str, Any]:
        treasury = {}
        fees_24h = {}
        now = int(time.time() * 1000)

        for asset in AssetSymbol:
            vault_acc = self.get_or_create_account(AccountType.EXCHANGE_VAULT, asset)
            fee_acc = self.get_or_create_account(AccountType.FEE_INCOME, asset)
            vault_dec = to_decimal(vault_acc["balance"])
            fee_dec = to_decimal(fee_acc["balance"])

            avail_dec = to_decimal(0)
            locked_dec = to_decimal(0)
            p2p_dec = to_decimal(0)
            for acc in db.accounts.values():
                if acc["asset"] == asset:
                    if acc["type"] == AccountType.USER_AVAILABLE:
                        avail_dec += to_decimal(acc["balance"])
                    elif acc["type"] == AccountType.USER_LOCKED:
                        locked_dec += to_decimal(acc["balance"])
                    elif acc["type"] == AccountType.P2P_ESCROW:
                        p2p_dec += to_decimal(acc["balance"])

            total_liab = avail_dec + locked_dec + p2p_dec
            is_solvent = vault_dec >= total_liab
            ratio_pct = ((vault_dec / total_liab) * 100) if total_liab > 0 else to_decimal(100)

            # Cold storage partition simulation (80% cold, 20% hot in vault)
            hot_dec = vault_dec * to_decimal("0.25")
            cold_dec = vault_dec - hot_dec

            treasury[asset.value] = {
                "hotWallet": format_decimal(hot_dec),
                "coldStorage": format_decimal(cold_dec),
                "totalExchangeAssets": format_decimal(vault_dec),
                "liabilities": {
                    "available": format_decimal(avail_dec),
                    "locked": format_decimal(locked_dec),
                    "pendingWithdrawal": "0.00000000",
                    "p2pEscrow": format_decimal(p2p_dec),
                    "total": format_decimal(total_liab)
                },
                "revenue": {
                    "tradingFees": format_decimal(fee_dec),
                    "withdrawalFees": "0.00000000"
                },
                "reserveRatio": f"{format_decimal(ratio_pct, 2)}%",
                "isSolvent": is_solvent,
                "withdrawalCapacityRatio": f"{format_decimal(ratio_pct, 1)}%"
            }

            fees_24h[asset.value] = {
                "baseFees": format_decimal(fee_dec),
                "quoteFees": "0.00000000"
            }

        return {
            "success": True,
            "treasury": treasury,
            "fees24h": fees_24h,
            "timestamp": now
        }

    def search_journal_entries(self, user_id: Optional[str] = None, asset: Optional[str] = None, account_id: Optional[str] = None, limit: int = 50, offset: int = 0) -> List[Dict[str, Any]]:
        entries = []
        for tx_id, tx in db.ledger_transactions.items():
            tx_entries = tx.get("entries", [])
            for e in tx_entries:
                acc_id = e.get("account_id", "")
                e_asset = e.get("asset")
                e_asset_str = e_asset.value if hasattr(e_asset, "value") else str(e_asset)

                if user_id and f"acc_{user_id}_" not in acc_id:
                    continue
                if asset and e_asset_str.upper() != asset.upper():
                    continue
                if account_id and acc_id != account_id:
                    continue

                entries.append({
                    "transaction_id": tx_id,
                    "idempotency_key": tx.get("idempotency_key"),
                    "description": tx.get("description"),
                    "account_id": acc_id,
                    "asset": e_asset_str,
                    "debit": e.get("debit", "0.00000000"),
                    "credit": e.get("credit", "0.00000000"),
                    "created_at": tx.get("created_at")
                })

        entries.sort(key=lambda x: x.get("created_at", 0), reverse=True)
        return entries[offset:offset + limit]


ledger_service = LedgerService()

